### ISSUE TITLE
byte-order-mark problem

### ISSUE THREAD

**Issue by dankelley** [dankelley/oce#1927]
I am seeing an error on the `r-devel-linux-x86_64-debian-clang` CRAN build (see https://cran.r-project.org/web/checks/check_results_oce.html and https://artifacts.r-hub.io/oce_1.7-3.tar.gz-28694b0caee744b88dde47ca14c293c0/oce.Rcheck/00check.log) so I installed docker to try to reproduce the bug.

*Proviso.* I have some notes on setting up docker at https://github.com/dankelley/oce/wiki/using-docker although I have to admit that (a) I have not been taking notes that are sufficient to reproduce my setup and (b) I am running docker with the `--rm` option (suggested in online sources) so I fear that when I end that session, I'll be stuck with the task of rebuilding the system for another test.  So, I'm planning to keep that docker session alive for a few days.

Anyway, after a few hours of twiddling, I can now build oce with R-devel on debian linux with clang.  And, lo and behold, I see an error.  Finally!  This should mean that I can start to tackle the problems with this platform.  (I am not entirely sure why just this one CRAN platform has the problem. Is it the clang version? The linux disto? The encoding?)
```R
> # Example 1: "csv1" Environment Canada format (found to be obsolete as of Oct 2019)
> csv1 <- read.met(system.file("extdata", "test_met_vsn1.csv", package="oce"))
Warning in readLines(file, warn = FALSE) :
  invalid input found on input connection '/oce.Rcheck/oce/extdata/test_met_vsn1.csv'
Error in read.table(file = file, header = header, sep = sep, quote = quote,  :
  incomplete final line found by readTableHeader on 'text'
Calls: read.met -> read.met.csv1 -> read.csv -> read.table
```

**Comment by dankelley** [comment]
Hm.  Somehow, `readLines()` is not getting the same number of lines as `wc` gets. How can that be?  (NOTE: the following comes from local tests in my docker, and it has extra I/O that is *not* in github. Thus, what I show below is not easily repeated by co-developers ... I am in "diary mode" here.)
```R
> source("oce/R/met.R");met1 <- read.met.csv1(csv1,debug=3)
  read.met.csv1("/opt/R-devel/lib/R/library/oce/extdata/test_met_vsn1.csv") {
    file has 17 lines
about to try to read
Error: $ operator is invalid for atomic vectors
>
[1]+  Stopped                 opt/R-devel/bin/R  (wd: /)
(wd now: /oce)
root@c8404c907079:/oce# wc /opt/R-devel/lib/R/library/oce/extdata/test_met_vsn1.csv
  23  110 1532 /opt/R-devel/lib/R/library/oce/extdata/test_met_vsn1.csv
```

**Comment by dankelley** [comment]
I don't get this because when I do as below in R on the debian image, I get the right result.  So something is wrong in my code... am I maybe using a 'file' that is open from a previous call, and that has been partly read???
```
f=file("/opt/R-devel/lib/R/library/oce/extdata/test_met_vsn1.csv", encoding="UTF-8-BOM")
length(readLines(f,warn=FALSE))
#[1] 23
```

**Comment by dankelley** [comment]
Hm. Below is what I get for the last line.
```
[17] "\"Date/Time\",\"Year\",\"Month\",\"Day\",\"Time\",\"Data Quality\",\"Temp ("
```
so, it stopped reading when it got to the "degree sign" that follows that parenthesis.

So, this is an encoding thing.  I guess I'll try some others.  But the only one I know that is said to handle BOM is `UTF-8-BOM`.

**Comment by dankelley** [comment]
Well, fumletck.  `encoding="latin1"` works on this test file.

I'll see if that will work on the other two test files.  It's test file 3 that has the BOM.

**Comment by dankelley** [comment]
Hm, `latin1` seems to work, even for the file with the BOM.  I'll need to fiddle around some more.  I think I just need to change the default encoding to `latin1`. Maybe that magically works for BOM.  Although I've had so much back-and-forth on this that I expect further bumps on the road.

That road will be more travelled this aft.

**Comment by dankelley** [comment]
Below (some output from tests) is a sign for the path ahead. Note that renaming is not working, which I think is because the strings must be coming out differently with the latin1 encoding.
```R
Error: sort(names(d@data)) (`actual`) not equal to sort(...) (`expected`).

     actual                 | expected                        
 [1] "Date/Time (LST)"      | "Date/Time (LST)" [1]           
 [2] "Dew Point Temp (Â°C)" - "dewPoint"        [2]           
 [3] "direction"            | "direction"       [3]           
 [4] "humidex"              | "humidex"         [4]           
 [5] "humidity"             | "humidity"        [5]           
 [6] "ï»¿Longitude (x)"     -                                 
 [7] "precipitation"        | "precipitation"   [6]           
 [8] "pressure"             | "pressure"        [7]           
 [9] "speed"                | "speed"           [8]           
[10] "Temp (Â°C)"           - "temperature"     [9]           
 ... ...                      ... 
> ```

**Comment by dankelley** [comment]
This is fixed in "develop" commit 1a0a83c0ffb79a2066273886f2a027eee34b6aa3.  To test it, I checked "develop" and "main" on (1) local linux builds and (2) rhub remote linux builds. The test was that "develop" would pass and "main" would fail, and this was the case for both local and remote.

Also, "develop" tests ok on remote windows.  (This is not a real test of this particular issue, because "main" also tests OK on remote windows.)

I have one more test: I'll use rhub to test for cran.  That takes several hours, so I'll post the results tomorrow.  If things fail with "main" and pass with "develop", then I'll have pretty high confidence that I've finished this task.  So, perhaps I'll close this tomorrow.

**Comment by dankelley** [comment]
These are the runs with "main".  I think one must be R-devel on one linux or another ... too bad the URLs don't indicate which is which.

```
> library(rhub)
> ?rhub::check_for_cran()
> rhub::check_for_cran(show_status=FALSE)
─  Building package
─  Uploading package
─  Preparing build, see status at
```

*   https://builder.r-hub.io/status/oce_1.7-3.tar.gz-b338c7111ce54a6d8b8903a5a47e2e4f
*   https://builder.r-hub.io/status/oce_1.7-3.tar.gz-ba6006f2b43c4d6cb13bd4c9d3e2341b
*   https://builder.r-hub.io/status/oce_1.7-3.tar.gz-34f91f393e3744c58fbdf423e5c13fe9
*  https://builder.r-hub.io/status/oce_1.7-3.tar.gz-e0ff6e4701bf4393820b0dbad87e7ab2

**Comment by dankelley** [comment]
Here are the runs with "develop".
```R
> rhub::check_for_cran(show_status=FALSE)
─  Building package
─  Uploading package
─  Preparing build, see status at
```
*   https://builder.r-hub.io/status/oce_1.7-3.tar.gz-72f11d9027f142d3aaa5fdcd3870620b
*   https://builder.r-hub.io/status/oce_1.7-3.tar.gz-5222547da7fc4ad5b99bbe211427a634
*   https://builder.r-hub.io/status/oce_1.7-3.tar.gz-fa679b3d6f81470c997b7d8bc7353a91
*   https://builder.r-hub.io/status/oce_1.7-3.tar.gz-eec303cf7e0849f98d816611b4409c44

**Comment by dankelley** [comment]
The `rhub::check_for_cran()` tests, shown in the last two comments, are a bit hard to interpret. First, the PREPERROR cases seem to be unrelated to oce itself, but rather stem from an inability to create a directory (for artifact files). Maybe that's a timeout thing, because oce is big and takes nearly 2 hours to test on these machines (which have to build up a whole R ecosystem from scratch).

What *really* matters are the previous tests I did, with local and remote tests on what I think is a close analogue to the CRAN debian R-devel clang system.  And these tests fail for "main" and pass for "develop", which is what I expected and hoped would happen.

Therefore I think I have solved this byte-order-mark problem.  As a summary, I recoded to
1. use `latin1` as the default encoding
2. improved the column-renaming code by making it more robust (I think) across encodings

Based on this, I am going to merge "develop" into "main", because "main" is certainly broken, and I don't see any benefit in leaving static at the CRAN code.  (PS I forgot to tag the submission, anyway, in the flurry of trying to get the thing past the first hurdle at CRAN.)

**Comment by dankelley** [comment]
I have merged "develop" into "main".  Next, I'll do
* [ ] local debian/r-devel/clang check
* [ ] remote debian/r-devel/clang 

I will "tick" here when they prove to be OK (if not OK, I'll make a further comment). Once (if) both are ticked, I'll close the issue.

**Comment by dankelley** [comment]
Hm.  The R-CMD-check action has an error on the windows machine. (The temporary link is https://github.com/dankelley/oce/runs/5718493075?check_suite_focus=true.)  It's related to the tests on `read_met()` (the example, actually).  I will investigate some more.  I did think at one point that windows was a special case, and from what I see on the development discussion group, it is.

Frankly, my inclination is to go to an extreme on this:
1. remove all tests
2. comment out the documentation examples

because even if I get it working at this moment, I worry that it might start failing, because my understanding is that this encoding business is in a state of flux, especially on windows.  Besides, do we really want this whole package failing CRAN checks because of some specialty functions?

**Comment by dankelley** [comment]
The test
```R
rhub::check(platform="windows-x86_64-release", show_status=FALSE)
```
yields (FILL IN LATER).

**Comment by dankelley** [comment]
The test
```R
devtools::check_win_release()
```
yields [FILL IN WHEN RUN COMPLETES].

**Comment by dankelley** [comment]
commit 4c33fbe2bffa440d3e58f579fa79f8de53b720e0
Merge: 76b7f786f f1ffc69b8
Author: dankelley <kelley.dan@gmail.com>
Date:   Mon Mar 28 13:16:32 2022 -0300

    Merge branch 'main' into develop

commit f1ffc69b86c4afc96e254e64a21f6222e1d7796b
Author: dankelley <kelley.dan@gmail.com>
Date:   Mon Mar 28 13:12:48 2022 -0300

    remove read.met examples
    
    They just are not worth the hassle they cause, in terms of checking on a
    variety of test machines that have different OSs, different versions of
    R, and different encodings.  The matrix of OS*Rversion*encoding is
    fairly large, and I don't know any easy way of testing them all.  The
    problem is that if any CRAN system (and I cannot replicate them
    properly, so far as I can tell) has a problem, then we get a big red
    'ERROR' message on the CRAN check page.  And, for what? So we can show
    something that ought to be obvious to anybody reading the man page.
    
    Note that there are also tests, and that I turn these off for Windows
    machines, which seem particularly prone to revealing problems.

**Comment by dankelley** [comment]
I've put the latest update ("develop" branch commit 4c33fbe2bffa440d3e58f579fa79f8de53b720e0) through a battery of tests, stated below.  A tick mark means that the test worked OK. If all get ticked, I'll close the issue.  Some of these tests take quite a while so that might be tomorrow morning.
1. [ ] local macos build & check

2. [ ] local debian build & check

3. [x] `rhub::check_for_cran()`

4. [x] `devtools::check_win_devel()`  

5. [ ] `rhub::check(platform="windows-x86_64-release", show_status=FALSE)`

6. [ ] `rhub::check(platform="debian-clang-devel",show_status=FALSE)`

**Comment by dankelley** [comment]
oce now builds and checks on a wide variety of platforms, and I think it will do the same on CRAN.  That's partly because I think I may have solved the problem, but mainly that I have finally come to agree with @richardsc that we should stop trying so hard ... I have dropped the reading of three sample files from the check sequence.  Those checks still get done if you build from github, but **not** if you build from source.  (This is because the `.Rbuildignore` file does not include a suite of developer tests, which use data stored in github but not supplied to the R package.)

I just wasted an hour trying to figure out a station number that will let us download met data for Halifax.  No luck.  And I was using `canadaHCD`, which is more sophisticated than oce (but which, I surmise, is using an index file that is hard-wired and thus out of date).

Basically, if a researcher needs to get data from EnvCan, it seems likely that they will know the station number, and how to download code.  And then, either `read.met()` will work or its docs will suggest a simple workaround, using `read.csv()` and `as.met()`.  With any luck, that researcher will let us know if `read.met()` failed, and then we will have a modern file format to work with.  This plan means that oce won't be subject to CRAN rejection for a minor, pimple-on-pimple, sort of feature.

So long, EnvCan, and thanks for all the phish.
